### PR TITLE
Prepare the Dart FFI package for publish

### PR DESCRIPTION
**Pull request by mosuem**
Prepare the Dart bindings for publishings.

Do not submit until https://github.com/truongsinh/icu-flutter/issues/3 is resolved.

### CODE REVIEW

File: ffi/capi/dart/package/LICENSE [hunk 1]
```diff
@@ -0,0 +1 @@
+--license TBD--
```
Comment (mosuem) [inline ffi/capi/dart/package/LICENSE:1]: What should the license be? @robertbastian @Manishearth

Comment (robertbastian) [reply ffi/capi/dart/package/LICENSE:1]: copy the license from the root folder

File: .github/workflows/publish-dart.yaml
Comment (robertbastian) [inline .github/workflows/publish-dart.yaml:1]: I don't think we want auto-publish in this repo. We're old-school.

File: ffi/capi/dart/package/pubspec.yaml [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
+version: 0.1.0
+description: Dart bindings for ICU4X.
+repository: https://github.com/unicode-org/icu4x/tree/main/ffi/capi/dart/package
 
 environment:
   sdk: '>=2.18.0'
 
 dependencies:
-  ffi: '>=1.1.0'
+  ffi: ^2.1.0
```
Comment (robertbastian) [inline ffi/capi/dart/package/pubspec.yaml:10]: why is this needed?

File: .github/workflows/publish-dart.yaml
Comment (mosuem) [reply .github/workflows/publish-dart.yaml:1]: You don't have to use the publishing feature, but the added comment on the PR showing if it can be published is still useful :)

Just as a data point: In the dart-lang repos, we are thinking about disabling manual publishing completely.

Comment (robertbastian) [reply .github/workflows/publish-dart.yaml:1]: What does it look like?

Comment (mosuem) [reply .github/workflows/publish-dart.yaml:1]: https://github.com/dart-lang/i18n/pull/736#issuecomment-1779362424

File: ffi/capi/dart/package/pubspec.yaml [hunk 3]
```diff
@@ -4,10 +4,10 @@ description: Dart bindings for ICU4X.
 repository: https://github.com/unicode-org/icu4x/tree/main/ffi/capi/dart/package
 
 environment:
-  sdk: '>=2.18.0'
+  sdk: '>=2.18.0 <4.0.0'
```
Comment (robertbastian) [inline ffi/capi/dart/package/pubspec.yaml:7]: does the linter check that this is compatible with each version? I remember seeing some errors about having to use 2.18 instead of 2.17, even though my sdk was 3.1.5.

File: ffi/capi/dart/package/pubspec.yaml [hunk 4]
```diff
@@ -4,10 +4,10 @@ description: Dart bindings for ICU4X.
...
 
 dependencies:
-  ffi: ^2.1.0
+  ffi: '>=1.1.0 <3.0.0'
```
Comment (robertbastian) [inline ffi/capi/dart/package/pubspec.yaml:10]: I don't actually know if this is true

File: .github/workflows/publish-dart.yaml
Comment (robertbastian) [reply .github/workflows/publish-dart.yaml:1]: If that happens to any change in `ffi/capi/dart`, that seems a bit annoying. We don't do out-of-cycle publishes, so I don't think there's much value in this.

File: ffi/capi/dart/package/pubspec.yaml (hunk 3 above)
Comment (mosuem) [reply ffi/capi/dart/package/pubspec.yaml:7]: Yes, the linter will complain when using features from a higher version than the pubspec constraint, independent of your installed SDK

File: ffi/capi/dart/package/pubspec.yaml (hunk 4 above)
Comment (mosuem) [reply ffi/capi/dart/package/pubspec.yaml:10]: You mean if versions `2.x` are fine to use?

File: .github/workflows/publish-dart.yaml
Comment (mosuem) [reply .github/workflows/publish-dart.yaml:1]: So no patches for fixes of the generated code? But I am also fine with removing this if it is too much noise.

Comment (robertbastian) [reply .github/workflows/publish-dart.yaml:1]: We haven't really had that situation yet, as no other language is published independently. We can do patches if needed, but I think this is too noisy.

File: ffi/capi/dart/package/pubspec.yaml (hunk 4 above)
Comment (robertbastian) [reply ffi/capi/dart/package/pubspec.yaml:10]: yes

Comment (robertbastian) [reply ffi/capi/dart/package/pubspec.yaml:10]: Yes, if the tests pick 2.x then 2.x works, if they pick 1.x then 1.x works, hard to tell

File: tools/make/tidy.toml
Comment (robertbastian) [inline tools/make/tidy.toml:1]: revert

File: ffi/capi/dart/package/pubspec.yaml (hunk 4 above)
Comment (mosuem) [reply ffi/capi/dart/package/pubspec.yaml:10]: I tested it with 2.x . In general, when giving a large range of version constraints, one should test using `dart pub upgrade` and `dart pub downgrade` to make sure both work. It's easier to just depend on the latest version to not have these issues, so `^2.1.0` in this case. Or is there a reason to stay with `1.x`?

Comment (robertbastian) [reply ffi/capi/dart/package/pubspec.yaml:10]: Ideally we'd restrict downstream clients as little as possible. I don't know how Dart dependency resolution works, but they might be in a situation where they can only use 1.x?

**Review by sffc — COMMENTED** [review]
This is for the Dart package? It looks fine but I trust you two on the details :)

File: ffi/capi/dart/package/README.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# Dart bindings for ICU4X
+[![Docs](https://docs.rs/icu/badge.svg)](https://docs.rs/icu) [![Build Status](https://github.com/unicode-org/icu4x/actions/workflows/build-test.yml/badge.svg)](https://github.com/unicode-org/icu4x/actions) [![pub package](https://img.shields.io/pub/v/icu.svg)](https://pub.dev/packages/icu) [![package publisher](https://img.shields.io/pub/publisher/icu.svg)](https://pub.dev/packages/icu/publisher)
+
+Experimental Dart bindings for ICU4X.
+
+Example:
+
+```dart
+main() {
```
Comment (robertbastian) [inline ffi/capi/dart/package/README.md:9]:
```suggestion
import `package:icu/icu.dart';

main() {
```

File: ffi/capi/dart/package/README.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# Dart bindings for ICU4X
+[![Docs](https://docs.rs/icu/badge.svg)](https://docs.rs/icu) [![Build Status](https://github.com/unicode-org/icu4x/actions/workflows/build-test.yml/badge.svg)](https://github.com/unicode-org/icu4x/actions) [![pub package](https://img.shields.io/pub/v/icu.svg)](https://pub.dev/packages/icu) [![package publisher](https://img.shields.io/pub/publisher/icu.svg)](https://pub.dev/packages/icu/publisher)
```
Comment (robertbastian) [inline ffi/capi/dart/package/README.md:2]: probably want to link to pub.dev docs instead of Rust?

**Review by robertbastian — APPROVED** [review]

**Comment by robertbastian** [comment]
I guess we can merge because it doesn't auto-publish?

**Comment by mosuem** [comment]
@robertbastian I would hold off until the name issue is resolved.

File: ffi/capi/dart/package/pubspec.yaml (hunk 4 above)
Comment (mosuem) [reply ffi/capi/dart/package/pubspec.yaml:10]: Our policy in general is update as soon as major version releases are available, as this shouldn't happen. I adapted the deps to reflect that.

**Comment by robertbastian** [comment]
Well the name is not changing with this PR.

**Comment by mosuem** [comment]
If the current owner of `package:icu` wants to keep it, we will have to change the name. I would prefer to do this in this PR, so that the package is indeed prepared for publishing :)

**Comment by JulianSchmidtZollsoft** [comment]
Hi :)
@mosuem do you have a rough estimate for when this dart package will be published? We want to use it for locale comparing Strings.

Thanks in advance.

**Comment by robertbastian** [comment]
We're at the moment undecided whether/how to release this package. Our goal is to release a new `intl` package, which uses ICU4X under the hood, this might or might not use this package. You can find the draft at https://github.com/dart-lang/i18n/tree/main/pkgs/intl, and there's a work-in-progress pull request to use ICU4X.

Open issues at the moment are binary size and how to do locale selection.

**Comment by robertbastian** [comment]
This is mostly obsolete
